pkg/template/processor.go

```diff
@@ -47,13 +49,54 @@ func apply(cl client.Client, objs []runtime.RawExtension) error {
 			continue
 		}
 		gvk := obj.GetObjectKind().GroupVersionKind()
-		if err := createOrUpdateObj(cl, obj); err != nil {
+		if err := createOrUpdateObj(p.cl, obj); err != nil {
 			return errs.Wrapf(err, "unable to create resource of kind: %s, version: %s", gvk.Kind, gvk.Version)
 		}
+		if gvk.Group == "project.openshift.io" && gvk.Version == "v1" && gvk.Kind == "ProjectRequest" {
```

Comment:
Other option is to keep this out of for loop.  First, we create project and wait for it to be active and then we run for_loop over other objects.  In this way, we can make sure that the template has ProjectRequest (which is mandatory) for given template.  Also, this if_cond no need to check in for_loop.

In any case, we should add logic that validate that given template has one project_request entry.

Reply:
well, keeping this inside the loop avoids code duplication (calling the `createOrUpdateObj` func and checking the error afterwards). 
Also, this whole processor could remain generic, and it's probably the responsibility of the caller to verify that the input template has a ProjectRequest. WDYT?

Reply:
For code duplication,
We can arrange code for least code duplication and it should be fine IMO.


For ProjectRequest check, 
I understand that, as a Template processor, it should not bother if `ProjectRequest` is present or mandatory .. but Template processor is only used by our product as of now then I think we can have this check somehow within Template processor else we need to first load this template one more time with `serializer.CodecFactory` and need to check.

Reply:
> For code duplication,
> We can arrange code for least code duplication and it should be fine IMO.
> 

The objects to create are sorted, so that the ProjectRequest will come first. Honestly, I don't see any benefit in processing the ProjectRequest separately from the other objects. The cost of checking if the object to create or update is a ProjectRequest is negligible.


> For ProjectRequest check,
> I understand that, as a Template processor, it should not bother if `ProjectRequest` is present or mandatory .. but Template processor is only used by our product as of now then I think we can have this check somehow within Template processor else we need to first load this template one more time with `serializer.CodecFactory` and need to check.

I'm not even sure what will be the benefit of checking if there's a ProjectRequest in the template. My point is: is there a chance that we will use this function without having a ProjectRequest object? Also, to avoid loading the template a second time, we could probably split the `ProcessAndApply` func to `Process` + `Apply`, so the caller can get the list of objects to create from the `Process` func and do some verifications before calling the `Apply` func.

Reply:
@nurali-techie I slightly refactored the code and there's no `ProcessAndApply` func anymore. You'll need to call `Process` first and get the `objects` to create (or an error), then pass those objects to the `Apply` func. So you can add some validation logic in between, to ensure that there's a `ProjectRequest` at the first position.